### DISCUSSION TITLE
Running tests on two multiple devices using react native

### DISCUSSION THREAD

**Discussion by tohel17** [appium/appium#16178] [Q&A] [answered]
I am using react native as my tech stack I want to run tests on multiple devices connected to a single appium server is it possible? if so how plz can anyone provide proper documentation for setup.

**Comment by dtopuzov** [comment, marked as answer]
Please check [Setup for Parallel Testing]( https://appium.io/docs/en/advanced-concepts/parallel-tests/) article.

**Reply by tohel17** [reply]
i saw this document but couldn't find any thing helpful can u please share anything else.

**Reply by dtopuzov** [reply]
It was very helpful for me, all capabilities you should pass in order to run in parallel are listed.

In general the approach is:
- Investigate how you can write parallel tests with testing framwork you use (Junit, TestNG, NUnit, Jest, ...)
- First make a prototype with parallel tests with out using appium at all
- Once you have that prove of concept working add appium and make sure you have correct capabilities

**Reply by tohel17** [reply]
thank you for sharing, may i know where to add those capabilities I mean in which file we need to add those.

**Reply by dtopuzov** [reply]
Steps are different based on programing language you want to use for your tests.
You should search for some getting started tutorial for appium with language of your choice.

**Reply by tohel17** [reply]
ok,thanks for your help.

**Reply by tohel17** [reply]
@dtopuzov I am still stuck on this issue, i declared capabilities for two devices, but I have doubt on how to initialize driver for both of them. to run the tests.
Here is my code snippet 
const config_device_1 = {
  platformName: "Android",
  platformVersion: "6.0.1",
    deviceName: "device",
    automationName: "UiAutomator2",
    appPackage: "************",
    appActivity: "*********,
    noReset: true,
    systemPort: 8200,
    udid: "***.*.**:5555",
    newCommandTimeout: "180000",//Appium will wait for a new command from the client before assuming the client quit and ending the session
  } //
  const config_device2 = {
    platformName: "Android",
    platformVersion: "6.0.1",
    deviceName: "device",
    automationName: "UiAutomator2",
   appPackage: "************",
    appActivity: "*********,
    noReset: true,
    systemPort: 8201,
    udid: "***.*.***:5555",
    newCommandTimeout: "180000",//Appium will wait for a new command from the client before assuming the client quit and ending the session
  }

//how to initialize driver for bith
beforeAll(async () => {
  await driver.init(config_device_1)
  await driver.sleep(50000)
})

**Comment by jlipps** [comment]
```
let driver1 = await driver.init(config_device_1)
let driver2 = await driver.init(config_device_2)
```

there you go! now you have two drivers.

**Reply by jlipps** [reply]
(this was supposed to be a reply to comment, not an answer to the OP)

**Reply by tohel17** [reply]
test("start and end the quick mode", async () => {
  //check if screen is loaded
  let TextId = await driver1.elementByAccessibilityId("welcometext")
  let TextId2 = await driver2.elementByAccessibilityId("welcometext")

  let text= await TextId.text()
  let text2= await TextId2.text()

  expect(text).toBe("Welcome")
  expect(text2).toBe("Welcome")
})


It means that we have to write seperate tests for every driver??

**Reply by jlipps** [reply]
that's still one test, is it not? if you have a single test that requires interaction between two devices, then yes, you need 2 driver objects. it's still one test.

if you're just trying to run multiple tests at the same time then you are going about it the wrong way. your code in that case should have just one driver object, and you need to use your test runner to parallelize test runs according to some rule. that's not something that is related to appium or that we can assist with, as it's a test runner feature. (you just need to make sure that when you are instantiating multiple drivers you are using the appropriate capabilities for each driver so they don't interfere with one another, i.e., they each pick a different device etc.)

**Reply by jlipps** [reply]
here's an article on how to achieve parallel testing in java: https://appiumpro.com/editions/28-running-multiple-appium-tests-in-parallel. with javascript, you'd need to refer to your test runner documentation to figure it out.

**Reply by tohel17** [reply]
Thanks for the help
